moshi/src/main/java/com/squareup/moshi/Types.java

```diff
@@ -49,6 +49,35 @@
   private Types() {
   }
 
+  /**
+   * Resolves the generated {@link JsonAdapter} fully qualified class name for a given
+   * {@link JsonClass JsonClass-annotated} {@code clazz}. This is the same lookup logic used by
+   * both the Moshi code generation as well as lookup for any JsonClass-annotated classes. This can
+   * be useful if generating your own JsonAdapters without using Moshi's first party code gen.
+   *
+   * @param clazz the class to calculate a generated JsonAdapter name for.
+   * @return the resolved fully qualified class name to the expected generated JsonAdapter class.
```

Comment:
Can you add a note that the returned class name will always be a top-level class, not a nested class? That way the interpretation of `.` is unambiguous.

Reply:
f9fd709